internal/olm/client/client.go

```diff
@@ -214,7 +216,53 @@ func (c Client) DoCSVWait(ctx context.Context, key types.NamespacedName) error {
 		return curPhase == olmapiv1alpha1.CSVPhaseSucceeded, nil
 	}
 
-	return wait.PollImmediateUntil(time.Second, csvPhaseSucceeded, ctx.Done())
+	err := wait.PollImmediateUntil(time.Second, csvPhaseSucceeded, ctx.Done())
+	if err != nil && errors.Is(err, context.DeadlineExceeded) {
+		depCheckErr := c.getDeploymentStatus(ctx, key, csv)
+		if depCheckErr != nil {
+			return fmt.Errorf("failed to run operator: %v %v", err, depCheckErr)
+		}
+	}
+
+	return err
+}
+
+// Get Deployment status
+
+func (c Client) getDeploymentStatus(ctx context.Context, key types.NamespacedName, csv olmapiv1alpha1.ClusterServiceVersion) error {
+	dep := &appsv1.Deployment{}
+	var depName string
+	for _, depSpec := range csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs {
+		if depSpec.Spec.Selector.MatchLabels["control-plane"] == "controller-manager" {
+			depName = depSpec.Name
+		}
+	}
+	depKey := types.NamespacedName{
+		Namespace: key.Namespace,
+		Name:      depName,
+	}
+	if depErr := c.KubeClient.Get(ctx, depKey, dep); depErr != nil {
+		return fmt.Errorf("failed to run operator: Error getting deployment: %v", depErr)
+	}
+	for _, s := range dep.Status.Conditions {
+		if s.Type == "Available" && s.Status == "True" {
+			return fmt.Errorf("failed to run operator: Deployment is available, but CSV pending")
+		}
+		if s.Type == "Progressing" || s.Type == "ReplicaFailure" {
+			// Loop through pods and return  error message.
+			podList := &coreV1.PodList{}
+			if podErr := c.KubeClient.List(ctx, podList); podErr != nil {
```

Comment:
This will return ALL pods, regardless of whether they're part of the Deployment, right?

I think we need to pass a ListOption to the List call using the deployments `spec.selector`.

Reply:
I did a quick test of sample operator. Both `pod` and `deployment` had common labels, in order to filter from both deployment spec list and pod list, as shown below.
```go
kind: Pod
metadata:
  creationTimestamp: "2020-09-03T14:27:36Z"
  generateName: memcached-operator-controller-manager-8687c65f7d-
  labels:
    control-plane: controller-manager
    pod-template-hash: 8687c65f7d
  name: memcached-operator-controller-manager-8687c65f7d-kc44t
  namespace: memcached-operator-system
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: memcached-operator-controller-manager-8687c65f7d
    uid: 3a3063eb-f167-4d41-96f1-f5cc49c723e9
  resourceVersion: "2722"
  selfLink: /api/v1/namespaces/memcached-operator-system/pods/memcached-operator-controller-manager-8687c65f7d-kc44t
  uid: d317bedc-7580-4abd-a566-ec80a437bb90
```
Deployment 
```go
  creationTimestamp: "2020-09-03T14:27:36Z"
  generation: 1
  labels:
    control-plane: controller-manager
  name: memcached-operator-controller-manager
  namespace: memcached-operator-system
  resourceVersion: "2697"
  selfLink: /apis/apps/v1/namespaces/memcached-operator-system/deployments/memcached-operator-controller-manager
  uid: 94a535c4-3324-4d10-883c-af326a092a56
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      control-plane: controller-manager
```